design/14_tinkerbell_crd_refactor/proposal.md

```diff
@@ -0,0 +1,462 @@
+
+	// TemplateRef is a reference to a Template resource used to render workflow actions.
+	// If no namespace is specified the Workflow's namespace is assumed.
+	TemplateRef LocalObjectReference
+
+	// TemplateData is user defined data that is injected during template rendering. The data
+	// structure should be marshalable.
+	// Optional.
+	TemplateData map[string]any
+
+	// Timeout defines the time the workflow has to complete. The timer begins when the first action
+	// is requested. When set to 0, no timeout is applied.
+	// Default 0.
+	Timeout int32
+}
+
+type WorkflowStatus struct {
+	// Actions is the list of rendered actions and their status.
+	Actions RenderedActions
+
+	// StartedAt is the time the first action was requested. Nil indicates the workflow has not
+	// started.
+	StartedAt *metav1.Time
+
+	// State describes the current state of the workflow. This fields represents a summation of
+	// action states. Specifically, if all actions succeeded, the workflow will succeed. If one
+	// action fails, the workflow fails irrespective of previous action status'.
+	State State
+
+	// Reason describes the reason for failure. It is only relevant when Result is ResultFailed.
+	// It is propogated from the failed action.
+	Reason Reason
+}
+
+// RenderedActions is a map of action name to RenderedAction.
+type RenderedActions map[ActionName]ActionStatus
+
+// ActionStatus describes status information about an action.
+type ActionStatus struct {
+	// Rendered is the rendered action.
+	Rendered Action
+
+	// StartedAt is the time the action was requested. Nil indicates the action has not started.
+	StartedAt *metav1.Time
+
+	// State describes the current state of the action.
+	State State
+
+	// Reason describes the reason for failure. It is only relevant when Result is ResultFailed.
+	Reason Reason
+
+	// Message is a freeform user friendly message describing the specific issue that caused the
+	// failure. It is only relevant when Result is ResultFailed.
+	Message string
+}
+
+// State describes the point in time state of a workflow or action.
+type State string
+
+const (
+	StatePending   State = "Pending"
+	StateRunning   State = "Running"
+	StateSucceeded State = "Succeeded"
+	StateFailed    State = "Failed"
+)
+
+// Reason is a one-word TitleCase string indicating why a failure occurred. It is not restricted
+// to the values defined with this API.
+type Reason string
+
+const (
+	ReasonUnknown Reason = "Unknown"
+	ReasonTimeout Reason = "Timeout"
+)
+```
+
+### Workflow state transition
+
+![Workflow state transitions](https://raw.githubusercontent.com/tinkerbell/roadmap/b7b2362997c01ffa52758e10259b8f55e81f7447/design/14_tinkerbell_crd_refactor/workflow_state_machine.png)
+
+Actions follow a similar state transition model.
+
+### Webhooks
+
+We will introduce a set of basic webhooks for validating each CRD. The webhooks will only validate the `v1alpha2` API version. They will be implemented as part of the `tink-controller` component.
```

Comment:
I know this might be controversial  😅 but can we please forget about webhooks and use CEL instead? Webhooks have been a huge pain to work with and debug and I would say this is a nice opportunity to integrate cel instead since this is still the beginning of the project. For more details about CEL PTAL at this https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/

Reply:
We will use CEL (I know its not clear in the proposal and I'd be happy to add specific annotations to clarify) but it doesn't capture all validation. For example, CEL won't be able to verify all `Hardware` submitted have unique `MAC` properties.